### PR TITLE
Adding profiling function

### PR DESCRIPTION
**Pull request by jychoi-hpc**
This is a PR for adding profiling routines based on PyTorch's Profiler.

### CODE REVIEW

File: hydragnn/train/train_validate_test.py [hunk 1]
```diff
@@ -35,8 +39,14 @@ def train_validate_test(
     verbosity=0,
     plot_init_solution=True,
     plot_hist_solution=False,
+    profile=False,
 ):
     num_epoch = config["Training"]["num_epoch"]
+    # setup profile variables
+    global profile_config_name
```
Comment (allaffa) [inline hydragnn/train/train_validate_test.py:46]: @jychoi-hpc is there any easy-fix to avoid the use of global variables?

**Review by allaffa — CHANGES_REQUESTED** [review]

File: hydragnn/train/train_validate_test.py (hunk 1 above)
Comment (jychoi-hpc) [reply hydragnn/train/train_validate_test.py:47]: Let me rewrite the code to remove global variables as much as possible.

**Comment by jychoi-hpc** [comment]
I have rewritten the code. I added a new class(`Profiler` in `hydragnn/utils/profile.py`) to remove global variables.

File: hydragnn/train/train_validate_test.py [hunk 2]
```diff
@@ -172,24 +186,38 @@ def get_head_indices(model, data):
...
-        loss, tasks_rmse = model.loss_rmse(pred, data.y, head_index)
-
-        loss.backward()
-        opt.step()
-        total_error += loss.item() * data.num_graphs
-        for itask in range(len(tasks_rmse)):
-            tasks_error[itask] += tasks_rmse[itask].item() * data.num_graphs
+    with profiler as prof:
+        for data in iterate_tqdm(loader, verbosity):
+            with record_function("load"):
```
Comment (streeve) [inline hydragnn/train/train_validate_test.py:205]: We'll have to rebase on main - we no longer load here

**Review by streeve — COMMENTED** [review]
This is a very nice added capability. I think I understand the basics, but is there any significant performance overhead (either with profiling turned on or off)?

It does make the training loop a little harder to read, but I think that's just the price we pay

File: hydragnn/train/train_validate_test.py (hunk 2 above)
Comment (jychoi-hpc) [reply hydragnn/train/train_validate_test.py:205]: Ok. Thank you for pointing this out. I am updating now.

File: hydragnn/train/train_validate_test.py [hunk 3]
```diff
@@ -172,24 +186,34 @@ def get_head_indices(model, data):
...
-
-        pred = model(data)
-        loss, tasks_rmse = model.loss_rmse(pred, data.y, head_index)
-
-        loss.backward()
-        opt.step()
-        total_error += loss.item() * data.num_graphs
-        for itask in range(len(tasks_rmse)):
-            tasks_error[itask] += tasks_rmse[itask].item() * data.num_graphs
+    with profiler as prof:
```
Comment (allaffa) [inline hydragnn/train/train_validate_test.py:201]: Is there a way to "hide" the action of the profiler so that someone who looks at the implementation of the training loop does not explicitly see the call tot he profile routines?

**Review by allaffa — CHANGES_REQUESTED** [review]

File: hydragnn/train/train_validate_test.py (hunk 3 above)
Comment (jychoi-hpc) [reply hydragnn/train/train_validate_test.py:201]: Alternatively, we can put the profiler wrapper in one higher up:
```
    for epoch in range(0, num_epoch):
        profiler.set_current_epoch(epoch)
        with profiler as prof:
            train_rmse, train_taskserr = train(
                train_loader, model, optimizer, verbosity, profiler=prof
            )
        val_rmse, val_taskserr = validate(val_loader, model, verbosity)
```
Then, the train function will look like this:
```
def train(...):
    tasks_error = np.zeros(model.num_heads)
    model.train()
    total_error = 0
    for data in iterate_tqdm(loader, verbosity):
        with record_function("zero_grad"):
            opt.zero_grad()
        with record_function("get_head_indices"):
            head_index = get_head_indices(model, data)
        with record_function("forward"):
            pred = model(data)
            loss, tasks_rmse = model.loss_rmse(pred, data.y, head_index)
        with record_function("backward"):
            loss.backward()
        opt.step()
        prof.step()
```

Comment (jychoi-hpc) [reply hydragnn/train/train_validate_test.py:201]: How do you think?

Comment (allaffa) [reply hydragnn/train/train_validate_test.py:201]: @streeve @pzhanggit 
Which option do you prefer between the one already implemented and the new one outlined by Jong in this conversation?

Comment (allaffa) [reply hydragnn/train/train_validate_test.py:201]: @jychoi-hpc 
I personally prefer the new proposed option. Let's see what Sam and Pei think :-)

Comment (streeve) [reply hydragnn/train/train_validate_test.py:201]: I agree - I think this is ready to merge after the update

Comment (jychoi-hpc) [reply hydragnn/train/train_validate_test.py]: I like the proposed option too. I have updated the code.

File: hydragnn/train/train_validate_test.py [hunk 4]
```diff
@@ -172,21 +188,30 @@ def get_head_indices(model, data):
...
-        pred = model(data)
-        loss, tasks_rmse = model.loss_rmse(pred, data.y, head_index)
-
-        loss.backward()
+        with record_function("zero_grad"):
+            opt.zero_grad()
+        with record_function("get_head_indices"):
+            head_index = get_head_indices(model, data)
+        with record_function("forward"):
+            pred = model(data)
```
Comment (streeve) [inline hydragnn/train/train_validate_test.py:209]: Should we separate the prediction from loss profiling?

**Review by streeve — APPROVED** [review]

**Review by allaffa — APPROVED** [review]

**Review by pzhanggit — APPROVED** [review]
